<a href="https://colab.research.google.com/github/cntrvsy/colab/blob/main/st20343345_CMP7005_PRAC1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# INTRODUCTION

- The serious health and environmental effects of air pollution, especially particulate matter like PM2.5, are the driving force behind this examination of Indian air quality data.  
- Finding "worst offender" cities and seasonal trends that might guide focused interventions requires an understanding of these pollutant levels and their total Air Quality Index (AQI).  
- In order to provide a thorough overview and predictive capability for air quality management, the analysis below attempts to find patterns in AQI over time, identify the most important pollutants leading to poor air quality, and predict AQI based on numerous criteria.

**Let's start by loading in our data from the GitHub Repo**

| No. | File Name                     | No. | File Name                     |
|-----|-------------------------------|-----|-------------------------------|
| 1   | Ahmedabad_data.csv            | 14  | Guwahati_data.csv             |
| 2   | Aizawl_data.csv               | 15  | Hyderabad_data.csv            |
| 3   | Amaravati_data.csv            | 16  | Jaipur_data.csv               |
| 4   | Amritsar_data.csv             | 17  | Jorapokhar_data.csv           |
| 5   | Bengaluru_data.csv            | 18  | Kochi_data.csv                |
| 6   | Bhopal_data.csv               | 19  | Kolkata_data.csv              |
| 7   | Brajrajnagar_data.csv         | 20  | Lucknow_data.csv              |
| 8   | Chandigarh_data.csv           | 21  | Mumbai_data.csv               |
| 9   | Chennai_data.csv              | 22  | Patna_data.csv                |
| 10  | Coimbatore_data.csv           | 23  | Shillong_data.csv             |
| 11  | Delhi_data.csv                | 24  | Talcher_data.csv              |
| 12  | Ernakulam_data.csv            | 25  | Thiruvananthapuram_data.csv   |
| 13  | Gurugram_data.csv             | 26  | Visakhapatnam_data.csv        |


In [ ]:
import pandas as pd
import numpy as np

city_names = [
    'Ahmedabad', 'Aizawl', 'Amaravati', 'Amritsar', 'Bengaluru', 'Bhopal',
    'Brajrajnagar', 'Chandigarh', 'Chennai', 'Coimbatore', 'Delhi', 'Ernakulam',
    'Gurugram', 'Guwahati', 'Hyderabad', 'Jaipur', 'Jorapokhar', 'Kochi',
    'Kolkata', 'Lucknow', 'Mumbai', 'Patna', 'Shillong', 'Talcher',
    'Thiruvananthapuram', 'Visakhapatnam'
]

dataframes = {}
for city in city_names:
    url = f'https://raw.githubusercontent.com/cntrvsy/colab/main/{city}_data.csv'
    dataframes[city] = pd.read_csv(url, header=0)


#FUNDERMENTAL DATA UNDERSTANDING

#What Is Our Data About?

our data consists of individual pollutants(PM2.5, PM10, NO, NO2, NOx, NH3, CO, SO2, O3, Benzene, Toluene, Xylene) collected over span of years. as a result we also have AQI column(Air Quality Index) that provides us with an overall grade which then we can infere the into a bucket such  as good or satifactory etc.

- **Why care about them?** While PM2.5 is a major concern, each of these other pollutants also has specific health impacts (e.g., CO affects oxygen transport, O3 causes respiratory issues, Benzene is carcinogenic) and environmental effects. They come from different sources (vehicles, industries, agriculture) and react with each other in the atmosphere. Understanding them individually helps us pinpoint pollution sources and design targeted solutions.

- **How is AQI calculated?** The AQI is typically determined by taking the worst (highest health risk) of the individual pollutant concentrations at a given time. If PM2.5 is high, it might drive the AQI to 'Poor'. If O3 is high, it could do the same. So, it's not a sum, but rather a representation of the dominant pollutant's impact on health.

In [ ]:
# combine it into one df
combined_df = pd.concat(dataframes.values())
combined_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 29531 entries, 0 to 1461
Data columns (total 16 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   City        29531 non-null  object 
 1   Date        29531 non-null  object 
 2   PM2.5       24933 non-null  float64
 3   PM10        18391 non-null  float64
 4   NO          25949 non-null  float64
 5   NO2         25946 non-null  float64
 6   NOx         25346 non-null  float64
 7   NH3         19203 non-null  float64
 8   CO          27472 non-null  float64
 9   SO2         25677 non-null  float64
 10  O3          25509 non-null  float64
 11  Benzene     23908 non-null  float64
 12  Toluene     21490 non-null  float64
 13  Xylene      11422 non-null  float64
 14  AQI         24850 non-null  float64
 15  AQI_Bucket  24850 non-null  object 
dtypes: float64(13), object(3)
memory usage: 3.8+ MB


In [ ]:
combined_df.head(3)


,City,Date,PM2.5,PM10,NO,NO2,NOx,NH3,CO,SO2,O3,Benzene,Toluene,Xylene,AQI,AQI_Bucket
0,Ahmedabad,01/01/2015,NaN,NaN,0.92,18.22,17.15,NaN,0.92,27.64,133.36,0.00,0.02,0.00,NaN,NaN
1,Ahmedabad,02/01/2015,NaN,NaN,0.97,15.69,16.46,NaN,0.97,24.55,34.06,3.68,5.50,3.77,NaN,NaN
2,Ahmedabad,03/01/2015,NaN,NaN,17.40,19.30,29.70,NaN,17.40,29.07,30.70,6.80,16.40,2.25,NaN,NaN


In [ ]:
#How many rows and columns are in your merged dataset?
combined_df.shape
print(f'No of Rows: {combined_df.shape[0]}, No of Columns: {combined_df.shape[1]}')

No of Rows: 29531, No of Columns: 16


In [ ]:
combined_df.columns

Index(['City', 'Date', 'PM2.5', 'PM10', 'NO', 'NO2', 'NOx', 'NH3', 'CO', 'SO2',
       'O3', 'Benzene', 'Toluene', 'Xylene', 'AQI', 'AQI_Bucket'],
      dtype='object')

In [ ]:
combined_df.describe()

,PM2.5,PM10,NO,NO2,NOx,NH3,CO,SO2,O3,Benzene,Toluene,Xylene,AQI
count,24933.000000,18391.000000,25949.000000,25946.000000,25346.000000,19203.000000,27472.000000,25677.000000,25509.000000,23908.000000,21490.000000,11422.000000,24850.000000
mean,67.450578,118.127103,17.574730,28.560659,32.309123,23.483476,2.248598,14.531977,34.491430,3.280840,8.700972,3.070128,166.463581
std,64.661449,90.605110,22.785846,24.474746,31.646011,25.684275,6.962884,18.133775,21.694928,15.811136,19.969164,6.323247,140.696585
min,0.040000,0.010000,0.020000,0.010000,0.000000,0.010000,0.000000,0.010000,0.010000,0.000000,0.000000,0.000000,13.000000
25%,28.820000,56.255000,5.630000,11.750000,12.820000,8.580000,0.510000,5.670000,18.860000,0.120000,0.600000,0.140000,81.000000
50%,48.570000,95.680000,9.890000,21.690000,23.520000,15.850000,0.890000,9.160000,30.840000,1.070000,2.970000,0.980000,118.000000
75%,80.590000,149.745000,19.950000,37.620000,40.127500,30.020000,1.450000,15.220000,45.570000,3.080000,9.150000,3.350000,208.000000
max,949.990000,1000.000000,390.680000,362.210000,467.630000,352.890000,175.810000,193.860000,257.730000,455.030000,454.850000,170.370000,2049.000000


### Inference from Descriptive Statistics

-  **Missing Values**: Many columns, particularly 'Xylene', 'NH3', and 'PM10', have significantly fewer `count` values than the total number of records, indicating a substantial presence of missing data that needs to be addressed.
-  **Skewness and Outliers**: For most pollutant concentrations (e.g., PM2.5, PM10, NO, CO) and the 'AQI', the `mean` is notably higher than the `50% (median)`. This suggests a right-skewed distribution, where pollutant levels are typically lower but occasionally spike to very high values. The large difference between the `75%` percentile and `max` values (e.g., AQI 75th percentile is 208, but max is 2049) further reinforces the presence of extreme outliers or severe pollution events.
- **Wide Range of Values**: The `min` and `max` values for pollutants and AQI span a very broad range, highlighting the high variability in air quality conditions across the collected data. This wide range implies diverse environmental conditions and pollution sources.
-  **Overall AQI**: The average AQI of approximately 166.46 (median 118.00) suggests that the air quality in the observed regions is frequently in the 'Moderate' to 'Poor' categories, with instances reaching 'Severe' levels as indicated by the maximum AQI. This underscores the importance of the analysis to identify contributing factors and patterns.

In [ ]:
#Which Pollutants are included, and which have the most missing data?
combined_df.isnull().sum()

,0
City,0
Date,0
PM2.5,4598
PM10,11140
NO,3582
NO2,3585
NOx,4185
NH3,10328
CO,2059
SO2,3854


In [ ]:
#How many unique cities are there?
combined_df['City'].nunique()

26

In [ ]:
def missing_values_table(combined_df):
    # Total missing values
    mis_val = combined_df.isnull().sum()

    # Percentage of missing values
    mis_val_percent = 100 * mis_val / len(combined_df)

    mis_val_table = pd.concat([mis_val, mis_val_percent], axis=1)

    mis_val_table = mis_val_table.rename(columns={0: 'Missing Values', 1: '% of Total Values'})

    # Sort the table by percentage of missing descending
    mis_val_table = mis_val_table.sort_values('% of Total Values', ascending=False)

    return mis_val_table

missing_values = missing_values_table(combined_df)
display(missing_values.style.background_gradient(cmap='Reds'))

,Missing Values,% of Total Values
Xylene,18109,61.322001
PM10,11140,37.723071
NH3,10328,34.973418
Toluene,8041,27.229014
Benzene,5623,19.041008
AQI,4681,15.851139
AQI_Bucket,4681,15.851139
PM2.5,4598,15.570079
NOx,4185,14.171549
O3,4022,13.619586


#DATA CLEANING
 - For most of the pollutant columns (PM2.5, NO, NO2, NOx, NH3, CO, SO2, O3, Benzene, Toluene, and AQI), the median is likely the most appropriate choice as Pollutant levels are rarely perfectly normally distributed and often have higher frequencies at lower concentrations with occasional high peaks.
 -Mean wouldnt work as it is sensitive to extreme values in the dataset.

 - we will add a "month" and "season" column as well for more data anylsis.

 ### AQI Bucket
 - to fill in this information we created a function to correctly map from the values from AQI.

In [ ]:
combined_df['Date'] = pd.to_datetime(combined_df['Date'], format="mixed", errors="coerce")

# adding a month column
combined_df['Month'] = combined_df['Date'].dt.month

# adding a seasons column
combined_df['Season'] = combined_df['Month'].apply(lambda x: 'Winter' if x in [12, 1, 2] else (
    'Spring' if x in [3, 4, 5] else (
    'Summer' if x in [6, 7, 8] else (
    'Autumn' if x in [9, 10, 11] else 'Unknown'))))


In [ ]:
# Dealing with missing values
# Xylene
xylene_median = combined_df['Xylene'].median()
combined_df['Xylene'] = combined_df['Xylene'].fillna(xylene_median)

#PM10
pm10_median = combined_df['PM10'].median()
combined_df['PM10'] = combined_df['PM10'].fillna(pm10_median)

#NH3
nh3_median = combined_df['NH3'].median()
combined_df['NH3'] = combined_df['NH3'].fillna(nh3_median)

#Toluene
toluene_median = combined_df['Toluene'].median()
combined_df['Toluene'] = combined_df['Toluene'].fillna(toluene_median)

#Benzene
benzene_median = combined_df['Benzene'].median()
combined_df['Benzene'] = combined_df['Benzene'].fillna(benzene_median)

#AQI
aqi_median = combined_df['AQI'].median()
combined_df['AQI'] = combined_df['AQI'].fillna(aqi_median)

#PM2.5
pm25_median = combined_df['PM2.5'].median()
combined_df['PM2.5'] = combined_df['PM2.5'].fillna(pm25_median)

#NOx
nox_median = combined_df['NOx'].median()
combined_df['NOx'] = combined_df['NOx'].fillna(nox_median)

#O3
o3_median = combined_df['O3'].median()
combined_df['O3'] = combined_df['O3'].fillna(o3_median)

#SO2
so2_median = combined_df['SO2'].median()
combined_df['SO2'] = combined_df['SO2'].fillna(so2_median)

#NO2
no2_median = combined_df['NO2'].median()
combined_df['NO2'] = combined_df['NO2'].fillna(no2_median)

#NO
no_median = combined_df['NO'].median()
combined_df['NO'] = combined_df['NO'].fillna(no_median)

#CO
co_median = combined_df['CO'].median()
combined_df['CO'] = combined_df['CO'].fillna(co_median)


In [ ]:
combined_df.isnull().sum()

,0
City,0
Date,0
PM2.5,0
PM10,0
NO,0
NO2,0
NOx,0
NH3,0
CO,0
SO2,0


## dealing with AQI Bucket
- We need a function to calculate AQI Bucket. it should be noted that different countries calculate this differently but for india it is as so:

| AQI Range | Category     | Health Advisory                                                                 |
|-----------|--------------|----------------------------------------------------------------------------------|
| 0–50      | Good         | None                                                                             |
| 51–100    | Satisfactory | Unusually sensitive people should reduce prolonged or heavy exertion            |
| 101–200   | Moderate     | Sensitive groups should reduce prolonged or heavy exertion                      |
| 201–300   | Poor         | Sensitive groups should avoid all outdoor activity; general public should limit prolonged or heavy exertion |
| 301–400   | Very Poor    | Sensitive groups should avoid all outdoor activity; general public should limit outdoor activity |
| 401–500   | Severe       | Everyone should avoid all physical activity outdoors                            |


In [ ]:
def aqi_to_bucket(aqi):
    if pd.isna(aqi):
        return np.nan
    aqi = float(aqi)
    if aqi <= 50:
        return "good"
    if aqi <= 100:
        return "satisfactory"
    if aqi <= 200:
        return "moderate"
    if aqi <= 300:
        return "poor"
    if aqi <= 400:
        return "very poor"
    return "severe"

combined_df['AQI_Bucket'] = combined_df['AQI'].apply(aqi_to_bucket)


In [ ]:
combined_df['AQI_Bucket'].value_counts()

,count
AQI_Bucket,
moderate,13510
satisfactory,8224
poor,2781
very poor,2337
good,1341
severe,1338


In [ ]:
missing_values_after_aqi_bucket = missing_values_table(combined_df)
display(missing_values_after_aqi_bucket.style.background_gradient(cmap='Reds'))

,Missing Values,% of Total Values
City,0,0.000000
Date,0,0.000000
PM2.5,0,0.000000
PM10,0,0.000000
NO,0,0.000000
NO2,0,0.000000
NOx,0,0.000000
NH3,0,0.000000
CO,0,0.000000
SO2,0,0.000000


# DATA OVERVIEW
now that we have dealt with the missing values in the dataset lets now have a general overview of the data

In [ ]:
# Lets import some extra libraries to help us visualize the data

!pip install "vegafusion[embed]>=1.5.0"
!pip install "vl-convert-python>=1.6.0"
!pip install streamlit

import altair as alt
import numpy as np

alt.renderers.enable('default')
alt.data_transformers.enable('vegafusion')

In [ ]:
# Temporarily rename 'PM2.5' to 'PM2_5' for charting compatibility
# Make a copy to avoid modifying the original DataFrame if not intended for global change
chart_df = combined_df.copy()
if 'PM2.5' in chart_df.columns:
    chart_df = chart_df.rename(columns={'PM2.5': 'PM2_5'})

# 1. Distribution of AQI_Bucket (horizontal)
aqi_bucket_chart = alt.Chart(chart_df).mark_bar().encode(
    y=alt.Y('AQI_Bucket:N', sort='-x', title='AQI Category'),
    x=alt.X('count():Q', title='Number of Readings'),
    tooltip=['AQI_Bucket', 'count()'],
    color=alt.Color('AQI_Bucket:N', legend=alt.Legend(title='AQI Category'))
).properties(
    title='Distribution of Air Quality Index (AQI) Buckets'
).interactive()

aqi_bucket_chart



# Inference
The chart above shows the frequency of different air quality categories,
indicating which categories are most common in the dataset.


In [ ]:
# Select only numeric columns
numeric_df = combined_df.select_dtypes(include=np.number)

# Exclude 'Month' and 'AQI'
# AQI will be included just to showcase the avg pollutant.
cols_to_analyze = [col for col in numeric_df.columns if col != 'Month']

descriptive_stats = numeric_df[cols_to_analyze].describe().T.reset_index()
descriptive_stats = descriptive_stats.rename(columns={'index': 'Pollutant'})
# Ensure column names are strings for Altair compatibility
descriptive_stats.columns = [str(col) for col in descriptive_stats.columns]

# Base chart for all pollutants
base = alt.Chart(descriptive_stats).encode(
    y=alt.Y('Pollutant:N', sort=alt.EncodingSortField(field="mean", op="mean", order='descending'), title="Pollutant"),
    tooltip=[
        alt.Tooltip('Pollutant:N', title='Pollutant'),
        alt.Tooltip('count:Q', title='Count'),
        alt.Tooltip('mean:Q', title='Mean', format='.2f'),
        alt.Tooltip('std:Q', title='Std Dev', format='.2f'),
        alt.Tooltip('min:Q', title='Min', format='.2f'),
        alt.Tooltip('25%:Q', title='25th Percentile', format='.2f'),
        alt.Tooltip('50%:Q', title='Median (50th Percentile)', format='.2f'),
        alt.Tooltip('75%:Q', title='75th Percentile', format='.2f'),
        alt.Tooltip('max:Q', title='Max', format='.2f'),
    ]
)

# Whiskers (min to max)
whiskers = base.mark_rule().encode(
    x=alt.X('min:Q', title='Value'),
    x2='max:Q'
)

# Box (25th to 75th percentile)
box = base.mark_bar(size=15, opacity=0.7).encode(
    x='25%:Q',
    x2='75%:Q',
    color=alt.value("steelblue")
)

# Median (50th percentile)
median_line = base.mark_tick(color='black', thickness=2, size=15).encode(
    x='50%:Q'
)

# Mean (as a diamond point)
mean_point = base.mark_point(shape='diamond', size=100, color='red').encode(
    x='mean:Q'
)

# Combine all layers to create the box-and-whisker style chart
chart = (whiskers + box + median_line + mean_point).properties(
    title='Descriptive Statistics of Pollutants (Box-and-Whisker Style)'
).interactive()

chart

#Inference
The chart effectively visualizes the range, spread, and central tendency for each pollutant. It quickly highlights which pollutants have a wider distribution (larger min-max range or IQR), how their median and mean values compare, and helps in identifying potential outliers or skewed distributions, providing a comprehensive statistical overview of the air quality parameters.

In [ ]:
seasonal_city_distribution = (
    alt.Chart(chart_df)
    .mark_bar()
    .encode(
        y=alt.Y('Season:N', sort=['Winter', 'Spring', 'Summer', 'Autumn'], title='Season'),
        x=alt.X('count():Q', title='Number of Readings'),
        color=alt.Color('Season:N', legend=alt.Legend(title='Season')),
        tooltip=['Season', 'count()']
    )
    .properties(
        title='Distribution of Readings Across Seasons by City',
        width=150,
        height=150
    )
    .facet(
    facet=alt.Facet('City:N', header=alt.Header(titleOrient="bottom", labelOrient="bottom")),
    columns=4
    )
    .interactive()
)

seasonal_city_distribution


### Insights from the Seasonal Distribution of Readings Across Cities

This faceted bar chart provides a detailed view of how data collection or availability varies across seasons for each of the 26 cities. Key observations include:

1.  **Variability Across Cities**: There is significant variability in the number of readings collected per season from city to city. Some cities show a relatively even distribution across all four seasons, which is good but shows the discrepancy in collection between cities.
2.  **Dominant Seasons**: For many cities, Winter and Spring appear to have a higher number of readings compared to Summer and Autumn. This could indicate seasonal patterns in data collection efforts, environmental monitoring, or potentially even data availability.
3.  **Data Gaps**: Certain cities might have fewer readings in particular seasons, suggesting potential data gaps or periods when monitoring was less frequent. For example, some cities might have very few or no readings during the Summer months.
4.  **City-Specific Patterns**: Each city exhibits its own unique seasonal distribution pattern. For instance, some cities might have their highest readings in winter, possibly due to increased air pollution during colder months or specific monitoring schedules.


In [ ]:
pm25_distribution_chart = alt.layer(
    # Histogram
    alt.Chart(chart_df).mark_bar(opacity=0.3).encode(
        x=alt.X('PM2_5:Q', bin=alt.Bin(maxbins=60), title='PM2.5 Concentration'),
        y=alt.Y('count():Q', title='Frequency'),
        tooltip=[alt.Tooltip('count()', title='Count')]
    ),

    # KDE line
    alt.Chart(chart_df).transform_density(
        'PM2_5',
        as_=['PM2_5', 'density']
    ).mark_line(size=2).encode(
        x='PM2_5:Q',
        y='density:Q',
        tooltip=['PM2_5:Q', 'density:Q']
    ),

    # Rug plot
    alt.Chart(chart_df).mark_tick(color='red', size=10, opacity=0.4).encode(
        x='PM2_5:Q',
        tooltip=['PM2_5']
    )
).properties(
    title='PM2.5 Distribution — Histogram + KDE + Rug',
    width=600,
    height=300
).interactive()

pm25_distribution_chart

#y-axis represents the 'Frequency' (how many data points fall into a specific PM2.5 concentration range)


This visualization provides a comprehensive view of how PM2.5 concentrations are distributed. The histogram bars show the frequency of PM2.5 values within specific ranges. The Kernel Density Estimate (KDE) line smooths out the histogram, giving a continuous estimate of the probability density function, which helps to identify the shape and peaks of the distribution. Finally, the rug plot adds individual data points along the x-axis, giving you a raw sense of the data density at each specific PM2.5 value. Together, these elements help understand the typical levels, spread, and any potential skewness or multiple peaks in PM2.5 concentrations.

# small summary

1) **Distribution of Air Quality Index (AQI) Buckets:**
This chart shows the frequency of different air quality categories, indicating which categories are most common in the dataset. From this, you can quickly see the overall air quality trend.

2) **Distribution of Readings Across Seasons:** This chart reveals if the data is evenly distributed across seasons or if certain seasons have more recorded data. This can highlight seasonal patterns or data collection biases.

3) **Distribution of PM2.5 Concentrations:** This histogram illustrates the range and most common concentrations of PM2.5, a key pollutant, in the dataset. It helps understand the typical levels and any extreme concentrations of PM2.5.



PM2.5 is a key pollutant primarily because of its significant impact on human health and the environment.
- **Size Matters for Health:** PM2.5 refers to particulate matter with a diameter of 2.5 micrometers or less. Because these particles are so small, they can bypass the body's natural defense mechanisms and penetrate deep into the lungs, and some can even enter the bloodstream.
- **Health Impacts:** Exposure to high levels of PM2.5 is linked to a wide range of serious health problems, including:
  - Respiratory issues (asthma, bronchitis, reduced lung function)
  - Cardiovascular diseases (heart attacks, strokes)
  - Premature death
  - Exacerbation of existing lung and heart conditions
- **Sources:** PM2.5 can come from various sources, both natural and man-made, such as vehicle exhaust, industrial emissions, burning fossil fuels, wildfires, and certain chemical reactions in the atmosphere.
- **Environmental Impacts:** Beyond health, PM2.5 also contributes to reduced visibility (haze), acid rain, and can affect ecosystems by altering nutrient balances in soil and water.

# DATA ANALYSIS

### Historical Trend
How has the average Air Quality Index (AQI) in India changed over the years? Why: To identify if pollution is getting better or worse overall.

In [ ]:
chart1 = alt.Chart(combined_df).mark_line(point=True).encode(
    x=alt.X('yearmonth(Date):T', title='Date'),
    y=alt.Y('mean(AQI):Q', title='Average AQI'),
    tooltip=['yearmonth(Date)', 'mean(AQI)']
).properties(
    title='Overall AQI Trend Over Time (Aggregated Mean)',
    width=600,
    height=300
).interactive()

chart1.display()

# Inference:
'''
This chart illustrates the overall trend of average Air Quality Index (AQI) over time.
Observing the line, we can identify periods of higher or lower AQI,
indicating whether air quality has generally improved, worsened, or remained stable across the collected years.
Significant peaks or dips might correspond to specific environmental events, policy changes, or seasonal variations
that affect air pollution levels.
'''

### The "Worst Offenders"

Which 10 cities have the highest average PM2.5 levels (the most dangerous pollutant)? Why: To identify which geographic locations need the most urgent intervention.

In [ ]:
# Calculate the average PM2_5 per city
avg_pm25_per_city = chart_df.groupby('City')['PM2_5'].mean().reset_index()

# Sort by PM2_5 in descending order and get the top 10
top_10_cities_pm25 = avg_pm25_per_city.sort_values(by='PM2_5', ascending=False).head(10)

chart2 = alt.Chart(top_10_cities_pm25).mark_bar().encode(
    x=alt.X('PM2_5:Q', title='Average PM2.5 Level'),
    y=alt.Y('City:N', sort='-x', title='City'),
    color=alt.Color('PM2_5:Q', scale=alt.Scale(scheme='reds'), legend=alt.Legend(title='Average PM2.5')),
    tooltip=['City', alt.Tooltip('PM2_5:Q', title='Average PM2.5', format='.2f')]
).properties(
    title='Top 10 Cities with Highest Average PM2.5',
    width=600
)

chart2.display()

### Pollutant Correlation
Is there a correlation between PM2.5 and PM10 levels? Why: To see if these two particulate matters generally rise together. If they correlate strongly, solving the source of one might fix the other.

In [ ]:
chart3 = alt.Chart(chart_df.sample(5000)).mark_circle(size=60).encode(
    x=alt.X('PM2_5:Q', title='PM2.5 Concentration'),
    y=alt.Y('PM10:Q', title='PM10 Concentration'),
    color=alt.Color('AQI_Bucket:N', scale=alt.Scale(scheme='spectral')),
    tooltip=['City', 'Date', 'PM2_5', 'PM10', 'AQI_Bucket']
).properties(
    title='Correlation between PM2.5 and PM10',
    width=600,
    height=400
).interactive() # Allows panning and zooming

chart3.display()

# Inference:
'''The scatter plot visually demonstrates the relationship between PM2.5 and PM10 concentrations.
A clear positive correlation is observable, indicating that as PM2.5 levels increase,
PM10 levels generally tend to increase as well. The coloration by 'AQI_Bucket'
further reveals that higher concentrations of both pollutants often correspond
to 'poor', 'very poor', or 'severe' air quality categories, reinforcing their combined impact
on the overall Air Quality Index. This strong correlation suggests that sources contributing to
one type of particulate matter often contribute to the other, implying that interventions targeting
particulate pollution could address both simultaneously.
'''

### Seasonal Heatmap
How does pollution intensity change by month across different years? Why: To spot seasonal patterns (e.g., does pollution spike in winter months due to fog or crop burning?).

In [ ]:
chart4 = alt.Chart(combined_df).mark_rect().encode(
    x=alt.X('month(Date):O', title='Month'),
    y=alt.Y('year(Date):O', title='Year'),
    color=alt.Color('mean(AQI):Q', scale=alt.Scale(scheme='orangered'), title='Avg AQI'),
    tooltip=['year(Date)', 'month(Date)', 'mean(AQI)']
).properties(
    title='Heatmap of Average AQI by Month and Year'
)

chart4.display()

# Inference:
'''
This heatmap reveals seasonal and annual patterns in air quality.
Darker shades indicate higher average AQI (worse air quality), while lighter shades suggest better air quality.
It helps to identify specific months or years when pollution is consistently high or low,
pointing towards potential seasonal factors (like winter inversions, crop burning, or summer ozone formation)
or long-term annual trends in pollution management and climate.
'''

###Composition of Air Quality
What is the proportion of days categorized as 'Good', 'Severe', 'Poor', etc.? Why: To get a "Report Card" summary of the overall dataset.

In [ ]:
chart5 = alt.Chart(combined_df).mark_arc(innerRadius=50).encode(
    theta=alt.Theta('count():Q'),
    color=alt.Color('AQI_Bucket:N', title='Category'),
    tooltip=['AQI_Bucket', 'count()']
).properties(
    title='Distribution of AQI Buckets'
)

chart5.display()

# Inference:
'''This donut chart provides an immediate 'report card' of the overall air quality within the dataset by showing
the proportional distribution of different AQI categories (Good, Satisfactory, Moderate, Poor, Very Poor, Severe).
It quickly highlights which air quality levels are most prevalent,
giving a high-level overview of the environmental health status across all cities and dates included in the data.
'''

###Chemical Distribution Box Plot
How does the distribution of Nitrogen Dioxide (NO2) vary across different AQI Buckets? Why: To see if NO2 is a major contributor to "Severe" ratings or if it remains stable even when AQI is bad.

In [ ]:
chart6 = alt.Chart(combined_df).mark_boxplot(extent='min-max').encode(
    x=alt.X('AQI_Bucket:N', title='Air Quality Bucket'),
    y=alt.Y('NO2:Q', title='NO2 Concentration'),
    color='AQI_Bucket:N'
).properties(
    title='Distribution of NO2 Levels within AQI Buckets',
    width=600
)

chart6.display()

# Inference:
'''
This box plot visualizes how Nitrogen Dioxide (NO2) concentrations vary across different AQI categories.
It helps to determine if NO2 levels are significantly higher in 'Poor' or 'Severe'
AQI buckets compared to 'Good' or 'Satisfactory' ones. This can indicate whether NO2 is a major contributing
pollutant to overall poor air quality, or if its distribution remains relatively stable regardless of the overall AQI.
'''

### Small Interactive Dashboard

How do individual pollutants (NO, CO, SO2) compare for a specific City over time? Why: This uses a "Dropdown" interaction. You can select a city from a list to update the charts dynamically.

In [ ]:
# Create a selection dropdown
city_dropdown = alt.binding_select(options=sorted(combined_df['City'].unique()))
selection = alt.selection_point(fields=['City'], bind=city_dropdown, name='Select')

base = alt.Chart(combined_df).encode(
    x='Date:T'
).add_params(
    selection
).transform_filter(
    selection
)

# Line for NO
line_no = base.mark_line(color='red').encode(
    y=alt.Y('NO:Q', title='Nitric Oxide (NO)'),
    tooltip=['Date', 'NO']
).properties(height=150, width=600)

# Line for NO2
line_no2 = base.mark_line(color='blue').encode(
    y=alt.Y('NO2:Q', title='Nitrogen Dioxide (NO2)'),
    tooltip=['Date', 'NO2']
).properties(height=150, width=600)

#Line for NOx
line_nox = base.mark_line(color='green').encode(
    y=alt.Y('NOx:Q', title='Nitric Oxide (NOx)'),
    tooltip=['Date', 'NOx']
).properties(height=150, width=600)

#Line for NH3
line_nh3 = base.mark_line(color='orange').encode(
    y=alt.Y('NH3:Q', title='Ammonia (NH3)'),
    tooltip=['Date', 'NH3']
).properties(height=150, width=600)


# Line for CO
line_co = base.mark_line(color='purple').encode(
    y=alt.Y('CO:Q', title='Carbon Monoxide (CO)'),
    tooltip=['Date', 'CO']
).properties(height=150, width=600)

# Line for SO2
line_so2 = base.mark_line(color='yellow').encode(
    y=alt.Y('SO2:Q', title='Sulphur Dioxide (SO2)'),
    tooltip=['Date', 'SO2']
).properties(height=150, width=600)

#Line for O3
line_o3 = base.mark_line(color='pink').encode(
    y=alt.Y('O3:Q', title='Ozone (O3)'),
    tooltip=['Date', 'O3']
    ).properties(height=150, width=600)

#Line for Benzene
line_benzene = base.mark_line(color='brown').encode(
    y=alt.Y('Benzene:Q', title='Benzene'),
    tooltip=['Date', 'Benzene']
    ).properties(height=150, width=600)

#Line for Toluene
line_toluene = base.mark_line(color='gray').encode(
    y=alt.Y('Toluene:Q', title='Toluene'),
    tooltip=['Date', 'Toluene']
    ).properties(height=150, width=600)

#Line for Xylene
line_xylene = base.mark_line(color='black').encode(
    y=alt.Y('Xylene:Q', title='Xylene'),
    tooltip=['Date', 'Xylene']
    ).properties(height=150, width=600)



# Combine them vertically
dashboard = alt.vconcat(line_no, line_no2, line_nox, line_nh3,line_co, line_so2, line_o3, line_benzene, line_toluene,line_xylene, ).properties(
    title='Interactive City Analysis: Select a City from the dropdown below'
)

dashboard.display()

# Inference:
'''
This interactive dashboard allows for detailed time-series analysis of various individual pollutants
(NO, NO2, NOx, NH3, CO, SO2, O3, Benzene, Toluene, Xylene) for a selected city.
Users can choose a city from the dropdown to dynamically visualize how these pollutant
concentrations fluctuate over time. This tool is invaluable for identifying specific
pollutant trends within a city, understanding their temporal patterns,
and potentially correlating them with local events or conditions not captured in the general AQI.

#ML MODELS
We will be using Random Forest Regressor since our dataset is tabular in nature. this way we can
- **Handles Non-linearity**: It can capture complex, non-linear relationships between pollutants and the target.
- **Robustness**: Less prone to overfitting than a single decision tree and generally robust to outliers.
- **Feature Importance**: It can provide insights into which pollutants are most influential in predicting AQI.
Less Preprocessing: It doesn't require extensive feature scaling.

In [ ]:
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestRegressor
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score

##---Prepare the data---
# We'll use the 'chart_df' as it has 'PM2_5' (renamed from PM2.5) and imputed values.
# Exclude 'City', 'Date', 'Month', 'Season', 'AQI_Bucket' as features for AQI prediction.

# Ensure all feature columns are numeric. 'AQI' is the target.
feature_columns = ['PM2_5', 'PM10', 'NO', 'NO2', 'NOx', 'NH3', 'CO', 'SO2', 'O3', 'Benzene', 'Toluene', 'Xylene']
target_column = 'AQI'

X = chart_df[feature_columns]
y = chart_df[target_column]

# Verify no null values in X or y, as they should have been handled during cleaning
# If there are still any NaN values, we'll fill them with the median or drop rows, depending on context.
# For this demonstration, we assume they are filled as per previous steps.

# Split the data into training and testing sets (70% train, 30% test)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

print(f"Training set size: {X_train.shape[0]} samples")
print(f"Testing set size: {X_test.shape[0]} samples")


In [ ]:
# Initialize the Random Forest Regressor model
# n_estimators is the number of trees in the forest
# random_state for reproducibility
model = RandomForestRegressor(n_estimators=100, random_state=42, n_jobs=-1) # n_jobs=-1 uses all available processors

# Train the model on the training data
print("Training Random Forest Regressor...")
model.fit(X_train, y_train)
print("Model training complete.")


In [ ]:
##--Make predictions on the test set--
y_pred = model.predict(X_test)

# Evaluate the model's performance
mae = mean_absolute_error(y_test, y_pred)
mse = mean_squared_error(y_test, y_pred)
rmse = np.sqrt(mse)
r2 = r2_score(y_test, y_pred)

print("\nModel Evaluation:")
print(f"Mean Absolute Error (MAE): {mae:.2f}")
print(f"Mean Squared Error (MSE): {mse:.2f}")
print(f"Root Mean Squared Error (RMSE): {rmse:.2f}")
print(f"R-squared (R2): {r2:.2f}")

# You can also look at feature importance
feature_importances = pd.Series(model.feature_importances_, index=X.columns).sort_values(ascending=False)
print("\nFeature Importances:")
display(feature_importances)


In [ ]:
import altair as alt
import pandas as pd

# Ensure feature_importances is a DataFrame for Altair
feature_importance_df = feature_importances.reset_index()
feature_importance_df.columns = ['Feature', 'Importance']

# Create the Altair chart
chart_feature_importance = alt.Chart(feature_importance_df).mark_bar().encode(
    x=alt.X('Importance:Q', title='Feature Importance'),
    y=alt.Y('Feature:N', sort='-x', title='Pollutant Feature'),
    tooltip=[alt.Tooltip('Feature:N'), alt.Tooltip('Importance:Q', format='.4f')]
).properties(
    title='Feature Importance for AQI Prediction',
    width=600
).interactive()


chart_feature_importance.display()


## Inference

- The visualization clearly shows that PM2.5 and CO are by far the most influential pollutants in determining the overall AQI. Their bars are significantly longer than the others, indicating they have the largest impact on the model's predictions.
- This is a crucial insight: if you want to improve air quality or predict AQI accurately, monitoring and managing PM2.5 and CO levels are paramount. The other pollutants also contribute, but their individual impact is much smaller in comparison.

Model 2

In [ ]:
import pandas as pd

# Create a DataFrame with hypothetical new pollutant values for a single instance
# Make sure the column names match the features used for training (X.columns)
new_data_point = pd.DataFrame({
    'PM2_5': [55.0],    # Example PM2.5 value
    'PM10': [100.0],    # Example PM10 value
    'NO': [15.0],
    'NO2': [30.0],
    'NOx': [40.0],
    'NH3': [20.0],
    'CO': [1.5],
    'SO2': [10.0],
    'O3': [35.0],
    'Benzene': [2.0],
    'Toluene': [5.0],
    'Xylene': [1.0]
})

# Use the trained model to predict AQI for this new data point
predicted_aqi = model.predict(new_data_point)

print(f"Predicted AQI for the given pollutant levels: {predicted_aqi[0]:.2f}")


###Training of the Dataset

In [ ]:
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestRegressor
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score
import pandas as pd
from sklearn.preprocessing import OneHotEncoder

##---Prepare data for retraining---
# Ensure chart_df is available, as it's used to create X_extended
# Assuming chart_df has been defined and processed in previous steps

# Define the original features (pollutant concentrations)
original_feature_columns = ['PM2_5', 'PM10', 'NO', 'NO2', 'NOx', 'NH3', 'CO', 'SO2', 'O3', 'Benzene', 'Toluene', 'Xylene']

# Define additional features to include
categorical_features = ['City']
numerical_features_to_add = ['Month']

# Create a copy of chart_df for feature engineering
# Ensure 'PM2.5' is renamed to 'PM2_5' if not already done in chart_df for consistency
if 'PM2.5' in combined_df.columns:
    chart_df_processed = combined_df.rename(columns={'PM2.5': 'PM2_5'})
else:
    chart_df_processed = combined_df.copy()

# Combine original and new numerical features for X_base
X_base_numerical = chart_df_processed[original_feature_columns + numerical_features_to_add]

# One-hot encode the 'City' column
encoder = OneHotEncoder(handle_unknown='ignore', sparse_output=False)
encoded_cities = encoder.fit_transform(chart_df_processed[categorical_features])

# Create a DataFrame from the one-hot encoded cities
encoded_city_df = pd.DataFrame(encoded_cities, columns=encoder.get_feature_names_out(categorical_features), index=chart_df_processed.index)

# Concatenate all features to create the extended feature set
X_extended = pd.concat([X_base_numerical, encoded_city_df], axis=1)

# Define the target variable
y = chart_df_processed['AQI']

X_train_extended, X_test_extended, y_train_extended, y_test_extended = train_test_split(X_extended, y, test_size=0.3, random_state=42)

print(f"Training set size (extended features): {X_train_extended.shape[0]} samples")
print(f"Testing set size (extended features): {X_test_extended.shape[0]} samples")

In [ ]:
from sklearn.preprocessing import OneHotEncoder
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline

# Define the original features (pollutant concentrations)
original_feature_columns = ['PM2_5', 'PM10', 'NO', 'NO2', 'NOx', 'NH3', 'CO', 'SO2', 'O3', 'Benzene', 'Toluene', 'Xylene']

# Define additional features to include
categorical_features = ['City']
numerical_features_to_add = ['Month']

# Create a copy of chart_df for feature engineering
# Ensure 'PM2.5' is renamed to 'PM2_5' if not already done in chart_df for consistency
if 'PM2.5' in chart_df.columns:
    chart_df_processed = chart_df.rename(columns={'PM2.5': 'PM2_5'})
else:
    chart_df_processed = chart_df.copy()

# Combine original and new numerical features for X_base
X_base_numerical = chart_df_processed[original_feature_columns + numerical_features_to_add]

# One-hot encode the 'City' column
encoder = OneHotEncoder(handle_unknown='ignore', sparse_output=False)
encoded_cities = encoder.fit_transform(chart_df_processed[categorical_features])

# Create a DataFrame from the one-hot encoded cities
encoded_city_df = pd.DataFrame(encoded_cities, columns=encoder.get_feature_names_out(categorical_features), index=chart_df_processed.index)

# Concatenate all features to create the extended feature set
X_extended = pd.concat([X_base_numerical, encoded_city_df], axis=1)

# Define the target variable
y = chart_df_processed['AQI']

print("Shape of the original features (pollutants only):")
print(X.shape)
print("\nShape of the extended feature set (pollutants + month + one-hot encoded cities):")
print(X_extended.shape)

print("\nFirst 5 rows of the extended feature set:")
display(X_extended.head(3))


In [ ]:
# Initialize a new Random Forest Regressor model for extended features
model_extended = RandomForestRegressor(n_estimators=100, random_state=42, n_jobs=-1) # n_jobs=-1 uses all available processors

# Train the new model on the extended training data
print("Training Random Forest Regressor with extended features...")
model_extended.fit(X_train_extended, y_train_extended)
print("Model training with extended features complete.")


### Make predictions on the extended test set

In [ ]:
##--Make predictions on the extended test set--
y_pred_extended = model_extended.predict(X_test_extended)

# Evaluate the new model's performance
mae_extended = mean_absolute_error(y_test_extended, y_pred_extended)
mse_extended = mean_squared_error(y_test_extended, y_pred_extended)
rmse_extended = np.sqrt(mse_extended)
r2_extended = r2_score(y_test_extended, y_pred_extended)

print("\nModel Evaluation (with extended features):")
print(f"Mean Absolute Error (MAE): {mae_extended:.2f}")
print(f"Mean Squared Error (MSE): {mse_extended:.2f}")
print(f"Root Mean Squared Error (RMSE): {rmse_extended:.2f}")
print(f"R-squared (R2): {r2_extended:.2f}")

# Display feature importances for the extended model
feature_importances_extended = pd.Series(model_extended.feature_importances_, index=X_extended.columns).sort_values(ascending=False)
print("\nFeature Importances (extended model):")
display(feature_importances_extended)


##Inference
Model Evaluation (with extended features):

- Mean Absolute Error (MAE): 19.91 (This is a slight improvement from the previous model's MAE of 20.16)
- Mean Squared Error (MSE): 1767.63 (Also an improvement from 1814.13)
- Root Mean Squared Error (RMSE): 42.04 (Improved from 42.59)
R-squared (R2): 0.89 (Remained very strong, indicating 89% of AQI variance is explained by the features)

### **Feature Importances (extended model):**

Here are the top features based on their importance in the extended model:

- PM2_5: 0.463838 (Still the most influential feature)
- CO: 0.347308 (Remains the second most influential)
- PM10: 0.054819
- NO: 0.030664
- O3: 0.016714
- NO2: 0.015431
- NOx: 0.014528
- SO2: 0.014450
- Toluene: 0.009903
- Benzene: 0.008057
- Xylene: 0.006675
- Month: 0.004750 (A new, moderately important feature)
- NH3: 0.003363
- City_Ahmedabad: 0.003186 (The most important among city features)

This shows that PM2.5 and CO continue to be the primary drivers of AQI, but 'Month' and certain city-specific factors also play a noticeable, albeit smaller, role. The model is now more comprehensive and slightly more accurate!

In [ ]:
import altair as alt
import pandas as pd

# Ensure feature_importances_extended is a DataFrame for Altair
feature_importance_extended_df = feature_importances_extended.reset_index()
feature_importance_extended_df.columns = ['Feature', 'Importance']

# Create the Altair chart for extended features
chart_feature_importance_extended = alt.Chart(feature_importance_extended_df).mark_bar().encode(
    x=alt.X('Importance:Q', title='Feature Importance'),
    y=alt.Y('Feature:N', sort='-x', title='Feature'),
    tooltip=[alt.Tooltip('Feature:N'), alt.Tooltip('Importance:Q', format='.6f')]
).properties(
    title='Feature Importance for AQI Prediction (Extended Model)',
    width=600
).interactive()

chart_feature_importance_extended.display()


In [ ]:
import pandas as pd
import numpy as np

# --- Step 1: Define the input values for pollutants, month, and city ---
# You can change these values to test different scenarios
hypothetical_pollutants = {
    'PM2_5': 60.0,
    'PM10': 110.0,
    'NO': 20.0,
    'NO2': 35.0,
    'NOx': 45.0,
    'NH3': 25.0,
    'CO': 2.0,
    'SO2': 12.0,
    'O3': 40.0,
    'Benzene': 3.0,
    'Toluene': 6.0,
    'Xylene': 1.5
}
predicted_month = 1 # January
predicted_city = 'Ahmedabad'

# --- Step 2: Prepare the input DataFrame for the extended model ---
# Get the list of all features the extended model was trained with
# This comes from X_extended.columns, which is available in the kernel state
model_features = X_extended.columns.tolist()

# Create an empty DataFrame for one prediction instance
new_data_point_extended = pd.DataFrame(0.0, index=[0], columns=model_features)

# Populate pollutant values
for col, value in hypothetical_pollutants.items():
    if col in new_data_point_extended.columns:
        new_data_point_extended[col] = value

# Populate the 'Month' feature
new_data_point_extended['Month'] = predicted_month

# Populate the one-hot encoded 'City' feature
city_col_name = f'City_{predicted_city}'
if city_col_name in new_data_point_extended.columns:
    new_data_point_extended[city_col_name] = 1.0
else:
    print(f"Warning: City '{predicted_city}' not found in model's training data. Please choose from {list(X_extended.filter(like='City_').columns)}.")

# --- Step 3: Make the prediction ---
predicted_aqi_extended = model_extended.predict(new_data_point_extended)

print(f"Predicted AQI for {predicted_city} in Month {predicted_month} with given pollutant levels: {predicted_aqi_extended[0]:.2f}")

# Display the input data point to verify
print("\nInput data point for prediction:")
display(new_data_point_extended)


#APPLICATION CODE

In [ ]:
import joblib
import os

# Define the directory to save the model artifacts
model_dir = 'streamlit_artifacts'
os.makedirs(model_dir, exist_ok=True)

# Save the trained RandomForestRegressor model
joblib.dump(model_extended, os.path.join(model_dir, 'random_forest_model.pkl'))

# Save the OneHotEncoder for cities
joblib.dump(encoder, os.path.join(model_dir, 'one_hot_encoder.pkl'))

# Save the list of feature columns used by the extended model
joblib.dump(X_extended.columns.tolist(), os.path.join(model_dir, 'model_features.pkl'))

print(f"Model, encoder, and feature list saved to '{model_dir}/'")

In [ ]:
combined_df.to_csv('processed_data.csv', index=False)

In [ ]:
%%writefile utils.py
import pandas as pd
import streamlit as st

@st.cache_data
def load_data():
    """Load air quality data from GitHub repository"""
    try:
        # Use raw.githubusercontent.com for direct CSV access
        url = 'https://raw.githubusercontent.com/cntrvsy/colab/main/processed_data.csv'
        df = pd.read_csv(url)

        # Convert Date column to datetime if it exists
        if 'Date' in df.columns:
            df['Date'] = pd.to_datetime(df['Date'], errors='coerce')

        return df

    except Exception as e:
        st.error(f"Error loading data: {e}")
        st.info("Make sure processed_data.csv exists in your GitHub repository and is publicly accessible.")
        # Return empty DataFrame to prevent crashes
        return pd.DataFrame()


In [ ]:
%%writefile multiapp.py
import streamlit as st

class MultiApp:
    def __init__(self):
        self.apps = []

    def add_app(self, title, func):
        self.apps.append({
            "title": title,
            "function": func
        })

    def run(self):
        st.sidebar.title("Navigation")
        app = st.sidebar.radio(
            "Select Section",
            self.apps,
            format_func=lambda app: app["title"]
        )
        app["function"]()



In [ ]:
%%writefile app.py
import streamlit as st
from multiapp import MultiApp

import data_overview
import exploratory_analysis
import modelling_prediction
import conclusions

st.set_page_config(
    page_title="Air Quality Dashboard",
    layout="wide"
)

app = MultiApp()

app.add_app(" Data Overview", data_overview.app)
app.add_app("Exploratory Data Analysis", exploratory_analysis.app)
app.add_app("Modelling & Prediction", modelling_prediction.app)
app.add_app("🧠Conclusions", conclusions.app)

app.run()


In [ ]:
%%writefile data_overview.py
import streamlit as st
from utils import load_data

def app():
    st.title(" Data Overview")

    df = load_data()

    col1, col2, col3 = st.columns(3)
    col1.metric("Total Records", len(df))
    col2.metric("Cities", df["City"].nunique())
    col3.metric("Features", df.shape[1])

    st.subheader("Dataset Preview")
    st.dataframe(df.head(), use_container_width=True)

    st.subheader("Key Observations")
    st.markdown("""
    - Dataset spans multiple cities and pollutants
    - Contains temporal AQI measurements
    - Suitable for trend and predictive analysis
    """)


In [ ]:
%%writefile exploratory_analysis.py
import streamlit as st
import plotly.express as px
from utils import load_data

def app():
    st.title("Exploratory Data Analysis")

    df = load_data()

    tab1, tab2 = st.tabs(["AQI Trends", "Correlations"])

    with tab1:
        fig = px.line(
            df,
            x="Date",
            y="AQI",
            color="City",
            title="AQI Trends Over Time"
        )
        st.plotly_chart(fig, use_container_width=True)

    with tab2:
        st.subheader("Correlation Matrix")
        st.dataframe(df.corr(numeric_only=True))

    st.subheader("Inference")
    st.info(
        "Clear temporal and city-specific AQI patterns are observed, "
        "suggesting strong environmental and seasonal influences."
    )


In [ ]:
%%writefile modelling_prediction.py
import streamlit as st
import pandas as pd
import joblib
import os

# Define the directory where model artifacts are saved
MODEL_DIR = 'streamlit_artifacts'

# Load the trained model and encoder (using Streamlit's cache for efficiency)
@st.cache_resource
def load_model_artifacts():
    model_path = os.path.join(MODEL_DIR, 'random_forest_model.pkl')
    encoder_path = os.path.join(MODEL_DIR, 'one_hot_encoder.pkl')
    features_path = os.path.join(MODEL_DIR, 'model_features.pkl')

    model = joblib.load(model_path)
    encoder = joblib.load(encoder_path)
    model_features = joblib.load(features_path)
    return model, encoder, model_features

def app():
    st.title("Modelling and Prediction")

    # Load model and encoder
    model_extended, encoder, model_features = load_model_artifacts()

    st.subheader("Model Evaluation")
    st.markdown("Here's a summary of the extended model's performance:")
    st.table(pd.DataFrame({
        "Metric": ["Mean Absolute Error (MAE)", "Mean Squared Error (MSE)", "Root Mean Squared Error (RMSE)", "R-squared (R2)"],
        "Value": [f"{19.91:.2f}", f"{1767.63:.2f}", f"{42.04:.2f}", f"{0.89:.2f}"] # Using previously computed values
    }))
    st.info("The model with extended features (including Month and City) shows a slight improvement in performance compared to the pollutants-only model, explaining 89% of the variance in AQI.")

    st.subheader("Predict AQI for Custom Inputs")
    st.markdown("Adjust the pollutant levels, month, and city below to get an estimated AQI.")

    # Get unique cities from the encoder (assuming the encoder was fitted on the full dataset)
    all_cities = sorted(encoder.categories_[0].tolist())

    # Input widgets for prediction
    col1, col2, col3 = st.columns(3)
    with col1:
        st.write("**Pollutant Levels (μg/m³ unless specified)**")
        pm2_5 = st.slider('PM2.5', 0.0, 500.0, 55.0, 0.1)
        pm10 = st.slider('PM10', 0.0, 1000.0, 100.0, 0.1)
        no = st.slider('NO', 0.0, 400.0, 15.0, 0.1)
        no2 = st.slider('NO2', 0.0, 400.0, 30.0, 0.1)
        nox = st.slider('NOx', 0.0, 500.0, 40.0, 0.1)
    with col2:
        st.write("") # Spacer
        nh3 = st.slider('NH3', 0.0, 400.0, 20.0, 0.1)
        co = st.slider('CO (mg/m³)', 0.0, 200.0, 1.5, 0.1)
        so2 = st.slider('SO2', 0.0, 200.0, 10.0, 0.1)
        o3 = st.slider('O3', 0.0, 300.0, 35.0, 0.1)
        benzene = st.slider('Benzene', 0.0, 500.0, 2.0, 0.1)
        toluene = st.slider('Toluene', 0.0, 500.0, 5.0, 0.1)
    with col3:
        st.write("") # Spacer
        xylene = st.slider('Xylene', 0.0, 200.0, 1.0, 0.1)

        st.write("**Temporal & Location**")
        month = st.selectbox('Month', range(1, 13), index=0) # Jan = 1, Dec = 12
        city = st.selectbox('City', all_cities, index=all_cities.index('Ahmedabad') if 'Ahmedabad' in all_cities else 0)

    if st.button('Predict AQI'):
        # Create a DataFrame for the new data point, ensuring all columns match model_features
        new_data_point = pd.DataFrame(0.0, index=[0], columns=model_features)

        # Populate pollutant values
        new_data_point['PM2_5'] = pm2_5
        new_data_point['PM10'] = pm10
        new_data_point['NO'] = no
        new_data_point['NO2'] = no2
        new_data_point['NOx'] = nox
        new_data_point['NH3'] = nh3
        new_data_point['CO'] = co
        new_data_point['SO2'] = so2
        new_data_point['O3'] = o3
        new_data_point['Benzene'] = benzene
        new_data_point['Toluene'] = toluene
        new_data_point['Xylene'] = xylene

        # Populate month
        new_data_point['Month'] = month

        # Populate one-hot encoded city
        city_col_name = f'City_{city}'
        if city_col_name in new_data_point.columns:
            new_data_point[city_col_name] = 1.0
        else:
            st.error(f"Selected city '{city}' was not found in the model's training data features. Please select another city.")
            return

        # Make prediction
        try:
            predicted_aqi = model_extended.predict(new_data_point)[0]
            st.success(f"Predicted AQI for the given conditions: **{predicted_aqi:.2f}**")

            # Convert AQI to bucket for better interpretation
            def aqi_to_bucket(aqi_val):
                if aqi_val <= 50: return "Good"
                elif aqi_val <= 100: return "Satisfactory"
                elif aqi_val <= 200: return "Moderate"
                elif aqi_val <= 300: return "Poor"
                elif aqi_val <= 400: return "Very Poor"
                else: return "Severe"

            aqi_category = aqi_to_bucket(predicted_aqi)
            st.info(f"This corresponds to an AQI category of: **{aqi_category}**")
        except Exception as e:
            st.error(f"An error occurred during prediction: {e}")

In [ ]:
%%writefile conclusions.py
import streamlit as st

def app():
    st.title(" Conclusions")

    st.markdown("""
    ### Summary
    - AQI varies strongly across cities and seasons
    - Data preprocessing improves model reliability
    - Ensemble models provide the best predictive performance

    ### Future Enhancements
    - Incorporate meteorological variables
    - Enable real-time AQI prediction
    - Deploy dashboard publicly
    """)

    st.info("This dashboard reflects an end-to-end analytical workflow.")


In [ ]:
%%writefile requirements.txt
pandas
numpy
streamlit
altair
plotly
scikit-learn
joblib
vegafusion
vl-convert-python